Proposes running big GitHub Actions only when PRs are ready for review.

vcanales opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
The recent GitHub Actions service degradation incident surfaced some quick wins that we can implement in order to save some computing time:

1. Pull request automation action, which usually runs in less than a minute — [history of runs here](https://github.com/WordPress/gutenberg/actions/workflows/pull-request-automation.yml?query=is%3Acompleted), go back a few pages to see run times with GitHub Actions running normally — should be forced to timeout after a _reasonable_ amount of time. Setting the timeout to 5 minutes — arbitrarily, we might be able to go lower — should be enough to point out real problems.
2. Limit execution of the biggest time consumers to PRs that are not Drafts, and Ready for Review:
    - End 2 end tests ([Web](https://github.com/WordPress/gutenberg/actions/workflows/end2end-test.yml), [iOS](https://github.com/WordPress/gutenberg/actions/workflows/rnmobile-ios-runner.yml), and [Android](https://github.com/WordPress/gutenberg/actions/workflows/rnmobile-android-runner.yml).)
    - [Performance Tests.](https://github.com/WordPress/gutenberg/actions/workflows/performance.yml)

[Static analysis](https://github.com/WordPress/gutenberg/actions/workflows/static-checks.yml), [unit tests](https://github.com/WordPress/gutenberg/actions/workflows/unit-test.yml?query=is%3Acompleted), etc. take less time and we might benefit from setting timeouts for those as well.

### Other alternative

We could add more granularity by using [if conditions](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsif) in combination with [Events.](https://docs.github.com/en/actions/reference/events-that-trigger-workflows)

## How has this been tested?
- Push changes on a draft PR, confirm that restricted workflows don't run.
- Mark PR as ready for review, confirm that restricted workflows run — no need to push new changes.

## Screenshots

### Admin e2e tests and Performance Tests remain in "Expected" status while the PR is still a Draft.

<img width="820" alt="image" src="https://user-images.githubusercontent.com/1157901/119057083-d46d4d80-b999-11eb-80db-62bbc158ba9c.png">

### Mobile e2e tests don't show up on the list.

<img width="393" alt="image" src="https://user-images.githubusercontent.com/1157901/119057188-08e10980-b99a-11eb-982b-0ad2cd2776f0.png">

vcanales committed:
try running e2e test only when ready for review

vcanales committed:
run mobile tests only with relevant changes

vcanales committed:
timeout pull request automation after 10 minutes

vcanales committed:
don't run performance test on draft PRs

gziolo commented on .github/workflows/rnmobile-android-runner.yml:
```diff
@@ -2,6 +2,9 @@ name: React Native E2E Tests (Android)
 on:
     pull_request:
+        types: [review_requested, ready_for_review]
+        paths:
```
I don't think that we should use paths here and for iOS, because React Native editor shares code with Gutenberg web.

gziolo reviewed (commented):
That's a very nice idea. I agree that we should explore better handling for PRs in the draft mode. I think we can be more aggressive in preventing more workflows from running like:
- Compressed Size
- Build Gutenberg Plugin Zip

Is it possible to use labels as triggers so repository members could apply them manually to start additional workflows without converting PR from draft? I guess it's something less important that we can investigate later.

ocean90 commented on .github/workflows/end2end-test.yml:
```diff
@@ -2,6 +2,7 @@ name: End-to-End Tests
 
 on:
     pull_request:
+        types: [review_requested, ready_for_review]
```
With that the test wouldn't run for further pushes on the PR, for example after a review. You'd have to request a review again which may be a workaround but might get a bit spammy. 

Maybe add also `synchronize` but check for `github.event.pull_request.draft` like done [here](https://github.community/t/dont-run-actions-on-draft-pull-requests/16817/16)? This would not directly skip the workflow but still would be faster than running the full workflow.

vcanales replied:
I just tried setting this PR back to draft, pushed something new, and it doesn't seem to stop checks from re-running. A Request for Changes here would be appreciated in order to check that case!

> Maybe add also `synchronize` but check for `github.event.pull_request.draft` like done [here](https://github.community/t/dont-run-actions-on-draft-pull-requests/16817/16)? This would not directly skip the workflow but still would be faster than running the full workflow.

Checking for draft, in my opinion, is not far enough to save us run time, as most PRs are not Drafts (139 out of 687 as of this comment.) I'll add synchronize to keep testing!

vcanales committed:
wip

desrosj committed:
Limit when workflows run on forks (#32114)

bobbingwide committed:
Remove class= wrapper to already wrapped attributes. Issue #29209 (#29214)

desrosj committed:
Improvements to NPM package caching across workflows (#32458)

* Use a consistent cache key for NPM packages across all workflows.

Also, include the NodeJS version in the cache key in case different package versions are required.

* Update the `actions/cache` action to the latest version.

* Remove the strategy matrix from jobs with a single NodeJS version.

For some reason, GitHub Actions will attach matrix values to job names in parenthesis. Because required checks are configured by job name, these need to stay consistent.

desrosj commented on .github/workflows/pull-request-automation.yml:
```diff
@@ -6,6 +6,7 @@ name: Pull request automation
 
 jobs:
     pull-request-automation:
+        timeout-minutes: 10
```
Does this ever reach this limit? I'm not sure that I have ever seen this workflow take more than 10 seconds.

vcanales replied:
This would have been helpful during the last GitHub Actions outage; this process in particular was running for hours in some cases. Now, I'm not certain that GHA would actually timeout the process in the context of an outage, but I think it's worth a try.

vcanales replied:
An example of this happening; this particular job [ran for 1h 2m](https://github.com/WordPress/gutenberg/actions/runs/857494540), on the day of a [big GitHub Actions outage](https://www.githubstatus.com/incidents/m16jzl31gnqt).

desrosj replied:
Got it! Seems like a good idea. If GHA is having an outage, I wonder if that limit would even work? I guess it depends on the outage itself.

Another option could be to add the `concurrency` option to this workflow. That would just cancel all other preceding workflows for the same branch. I don't know how the code is structured within `packages/project-management-automation` though. So I can't say whether it's essential that it runs for every `push` event.

desrosj commented:
I'm not yet convinced that this is the best way to save time and reduce the chance for a backlog of jobs to occur.

In general, I am pro running workflows for each commit/push so long as they are scoped properly, and earlier runs are cancelled when appropriate (fixed in #32016). Since #32016 was merged, we have not experienced a backlog that has resulted in any significant disruption.

Some things I would like to try and compare results for before merging this:
- Splitting the E2E test workflow into more parallel jobs. Each of the 4 sets of tests seem to take between 9-20 minutes specifically on the "Run Test" step of the action.
- The performance testing does not take advantage of running tests in parallel. It looks like there are two groups of tests that take around 5-9 minutes on each branch tested, and then results are posted. It would be great if each branch could be run in parallel and the results were added as artifacts for a third job to use for calculating results.

vcanales committed:
try running performance jobs as matrix

ockham commented on .github/workflows/build-plugin-zip.yml:
```diff
@@ -18,53 +18,20 @@ concurrency:
     cancel-in-progress: true
 
 jobs:
-    compute-stable-branches:
-        name: Compute current and next stable release branches
-        runs-on: ubuntu-latest
-        outputs:
-            current_stable_branch: ${{ steps.get_branches.outputs.current_stable_branch }}
-            next_stable_branch: ${{ steps.get_branches.outputs.next_stable_branch }}
-
-        steps:
-            - name: Get current and next stable release branches
-              id: get_branches
-              run: |
-                  curl \
-                    -H "Accept: application/vnd.github.v3+json" \
-                    -o latest.json \
-                    "https://api.github.com/repos/${{ github.repository }}/releases/latest"
-                  LATEST_STABLE_TAG=$(jq --raw-output '.tag_name' latest.json)
-                  IFS='.' read LATEST_STABLE_MAJOR LATEST_STABLE_MINOR LATEST_STABLE_PATCH <<< "${LATEST_STABLE_TAG#v}"
-                  echo "::set-output name=current_stable_branch::release/${LATEST_STABLE_MAJOR}.${LATEST_STABLE_MINOR}"
-                  if [[ ${LATEST_STABLE_MINOR} == "9" ]]; then
-                    echo "::set-output name=next_stable_branch::release/$((LATEST_STABLE_MAJOR + 1)).0"
-                  else
-                    echo "::set-output name=next_stable_branch::release/${LATEST_STABLE_MAJOR}.$((LATEST_STABLE_MINOR + 1))"
-                  fi
-
```
Removing this would completely break our [release tooling](https://github.com/WordPress/gutenberg/blob/cba842a101df8a0cc0c2476768facc0c1e017532/docs/contributors/code/release.md#release-tool). (Same for a number of other changes below.) I'm guessing this wasn't deliberate but the result of a rebase that didn't go quite right? 😅

vcanales replied:
100% — this branch was quite behind, I'll put everything back 😅

vcanales commented:
Closing as stale, and fixed somewhere else.
